### PR TITLE
allow skip proactive cleaning segment lineages

### PR DESCRIPTION
**Pull request by klsince**
## Description
There are some use cases of SegmentLineage, where the segmentFrom sets are always empty as the output segments are not generated from other segments. `forceClean` would revert those lineages as their segmentFrom are all empty per [logic here](https://github.com/apache/pinot/blob/master/pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/PinotHelixResourceManager.java#L2888)

This PR allows method caller to decide whether to force clean. Without force cleaning, the lineages are cleaned as part of retention mgmt in background. 

## Upgrade Notes
Does this PR prevent a zero down-time upgrade? (Assume upgrade order: Controller, Broker, Server, Minion)
* [ ] Yes (Please label as **<code>backward-incompat</code>**, and complete the section below on Release Notes)

Does this PR fix a zero-downtime upgrade introduced earlier?
* [ ] Yes (Please label this as **<code>backward-incompat</code>**, and complete the section below on Release Notes)

Does this PR otherwise need attention when creating release notes? Things to consider:
- New configuration options
- Deprecation of configurations
- Signature changes to public methods/interfaces
- New plugins added or old plugins removed
* [ ] Yes (Please label this PR as **<code>release-notes</code>** and complete the section on Release Notes)

### CODE REVIEW

**Comment by jackjlli** [comment]
I think what you're trying to do here is to support consistent push for append use cases, e.g. for daily data push. If that's the case, the logic should handle the following scenarios:
* For refresh table, it's always not allowed to have two concurrent push happening at the same time, regardless of the emptiness of the existing table. In this case, the previous `IN_PROGRESS` segment lineage should be cleaned up.
* For append table, it's okay to have more than one batch push at the same time, as long as they're not pushing for the same set of data (i.e. they're not overwriting each other).

So going back to the code, I think it'd be good to include the validation above instead of just releasing the `forceCleanup` to false. cc: @snleee @jtao15

**Comment by jtao15** [comment]
Even if `forceClean = true`, we should allow creating multiple lineages with empty `SegmentsFrom` list. I think we can add one more condition `!segmentsFrom.isEmpty()` on `L2886` to handle this. Also, we need to allow creating lineage entries with the same empty `SegmentsFrom` list on `L2922`.

**Comment by jackjlli** [comment]
> Even if `forceClean = true`, we should allow creating multiple lineages with empty `SegmentsFrom` list. I think we can add one more condition `!segmentsFrom.isEmpty()` on `L2886` to handle this. Also, we need to allow creating lineage entries with the same empty `SegmentsFrom` list on `L2922`.

Just adding an empty list check isn't sufficient, there could be two batch jobs pushing for the same target set of data. E.g. for refresh table, it's always not allowed to have such two batch jobs running at the same time. For append table, if two batch jobs are pushing the same data, one of them should be rejected or stopped.

**Comment by jtao15** [comment]
> > Even if `forceClean = true`, we should allow creating multiple lineages with empty `SegmentsFrom` list. I think we can add one more condition `!segmentsFrom.isEmpty()` on `L2886` to handle this. Also, we need to allow creating lineage entries with the same empty `SegmentsFrom` list on `L2922`.
> 
> Just adding an empty list check isn't sufficient, there could be two batch jobs pushing for the same target set of data. E.g. for refresh table, it's always not allowed to have such two batch jobs running at the same time, even though the `segmentFrom` field is empty. For append table, if two batch jobs are pushing the same data, one of them should be rejected or stopped.

I thought creating two entries with empty `SegmentsFrom` will not happen for REFRESH table initially. But actually it can happen for empty tables. Agreed now.

**Comment by klsince** [comment]
hi @jackjlli @jtao15 thanks for the insights. I see currently conflict is checked with equality check like CollectionUtils.isEqualCollection. I wonder why it's not checking if there is overlap between segmentFrom sets instead? I'm thinking if I should check equality or overlap between segmentTo sets for the validation mentioned in your comments.

**Comment by jackjlli** [comment]
> hi @jackjlli @jtao15 thanks for the insights. I see currently conflict is checked with equality check like CollectionUtils.isEqualCollection. I wonder why it's not checking if there is overlap between segmentFrom sets instead? I'm thinking if I should check equality or overlap between segmentTo sets for the validation mentioned in your comments.

Even if `segmentsTo` sets don't have any overlapping, both `segmentsTo` fields could still point to the same time range. Think of the refresh table, how do we make sure two batches don't overwrite each other if the segment names from two batch jobs are the same.

**Comment by klsince** [comment]
I carved out a [shorter PR](https://github.com/apache/pinot/pull/8242) with more narrowed scope of change, and will follow up on comprehensive conflict checks in future PRs.

**Review by snleee — APPROVED** [review]

File: pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/PinotHelixResourceManager.java [hunk 1]
```diff
@@ -2878,13 +2878,13 @@ public String startReplaceSegments(String tableNameWithType, List<String> segmen
...
           // When 'forceCleanup' is enabled, we need to proactively clean up at the following cases:
-          // 1. Revert the lineage entry when we find the lineage entry with the same 'segmentFrom' values. This is
+          // 1. Revert the lineage entry when we find the lineage entry with overlapped 'segmentFrom' values. This is
           //    used to un-block the segment replacement protocol if the previous attempt failed in the middle.
           // 2. Proactively delete the oldest data snapshot to make sure that we only keep at most 2 data snapshots
           //    at any time in case of REFRESH use case.
           if (forceCleanup) {
-            if (lineageEntry.getState() == LineageEntryState.IN_PROGRESS && CollectionUtils
-                .isEqualCollection(segmentsFrom, lineageEntry.getSegmentsFrom())) {
+            if (lineageEntry.getState() == LineageEntryState.IN_PROGRESS && !Collections
```
Comment (klsince) [inline pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/PinotHelixResourceManager.java:2886]: found I pushed this change accidentally. I'll revert/refine this part in coming PR: https://github.com/apache/pinot/pull/8242 

but interestingly, looks like all tests were fine with this change, even disjoint is more strict than equality check.

**Pull request merged**